File: 39组合总和.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        size = len(candidates)

        def dfs(candidate, begin, size, path, res, target):
            if target < 0:
                return
            elif target == 0:
                res.append(path)
            for index in range(begin, size):
                dfs(candidate, index, size, path + [candidate[index]], res, target - candidate[index]) 

        res = []
        path = []

        dfs(candidates, 0, size, path, res, target)

        return res

File: 39组合总和.swift
class Solution {
    var result = [[Int]]()
    var res = [Int]()
    func combinationSum(_ candidates: [Int], _ target: Int) -> [[Int]] {
        backSearch(candidates, target, 0, 0)
        return result
    }

    func backSearch(_ candidates: [Int], _ target: Int, _ sum: Int, _ startIndex: Int) {
        if sum > target {
            return
        }
        if sum == target {
            result.append(res)
            return
        }
        if startIndex >= candidates.count {
            return
        }
        for index in startIndex..<candidates.count {
            res.append(candidates[index])
            backSearch(candidates, target, sum+candidates[index], index)
            res.popLast()
        }
    }
}
